### PR TITLE
DDP-4229 verify age up at registration

### PR DESCRIPTION
**Pull request by zyme**
## Context and the big picture
 Last piece of DDP-4229 is consulting the age up governance policy for the study when registering from an invitation and verifying that the user has aged up.
 
 ## What type of change is this?
  
  - [ ] Bugfix (non-breaking change which fixes an issue)
  - [x] New feature (non-breaking change which adds functionality)
  - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - [ ] Experimental proof-of-concept
  - [ ] Infosec/Compliance remediation
  
 ## Risk Assessment
  - [x] These changes are **HIGH RISK**.  Changes to the user registration route are pretty much always high risk, eh?
    - [ ] There is an elevated risk that may impact a large number of features
    - [ ] There is an elevated risk to security and privacy
    - [ ] These changes introduce new 3rd party dependencies, alter the network perimeter, or contain
    major upgrades to persistent storage or SaaS dependencies such as auth0, sendgrid, or easypost    
  - [ ] These changes are low risk, do not impact security/privacy, do not impact pepper shared
  components/modules, and have no major upgrades to 3rd party services or libraries
 
 ### FUD Score 
 Overall, how are you feeling about these changes?
  - [x] :relaxed: All good, business as usual!
  - [ ] :sweat_smile: There might be some issues here
  - [ ] :scream:I'm sweaty and nervous
 
 ## How to we demo these changes?
 How does one observe these changes in a deployed system?  Note that _user visible_ encompasses
 many personas--not just patients and study staff, but ops your fellow devs, compliance, etc.
 
 - [ ] They are user-visible in dev as a regular user journey and require no additional instructions.
 - [ ] Getting dev into a state where this is user-visible requires some tech fiddling.  I have
 documented these steps in the related ticket.
 - [x] Requires other features before it's human visible.  These features are already in the sprint.
 - [ ] I have no idea how to demo this.  Please help me!
 
 ## UI/UX
  - [x] There ain't no UI/UX in these changes
  - [ ] My changes have passed muster with Erin and Jen via an over-the-shoulder review, screenshots, etc.
  - [ ] Erin and Jen have not had an opportunity to provide feedback yet
 
###  Backend
   - [x] I have considered error handling and notification to slack alert rooms via `LOG.error()`.
     #### Route Changes
     - [ ] My changes involve a new route changes to an existing route, so the who/what/when are logged using the usual logback mechanism
at the _start_ of the route, prior to any validation, as well as at the end of the route when work completes.
     ### Housekeeping Changes
     - [ ] My changes involve updates to housekeeping, and I am logging the who/what/when via logback.
  ### Client-side
   - [ ] My changes do not involve backend route change, but I have useful logging statements nonetheless.

## Analytics
 - [ ] I have discussed the analytics needs at both a platform and study level with Jen and instrumented code
 accordingly.
 - [x] There are no analytics requirements for these changes

## Security and Privacy
Ensuring proper security and privacy for our users is essential.

  ###  Backend
   - [ ] These changes alter fundamental auth filter logic
   - [ ] These changes introduce a new backend API route that is behind an auth filter
   - [ ] These changes introduce a new backend API that _does not require auth_
   - [ ] These changes alter auth0 internals (rules, hosted login, configuration, etc.)
   - [ ] These changes expose new information to elastic and I have verified that PHI is not being exposed to the public
    
  ### Browser
  - [ ] These changes alter what we store in browser storage
  
  ### Mobile
  - [ ] TBD 
 
 ## Performance and Stability
  - [ ] I have analyzed my changes for stability, fault tolerance, graceful degradation, and performance bottlenecks and written a brief summary in this PR
  - [ ] I'm unsure about stability, performance, fault tolerance, and graceful degradation.  Please help!
  
## Testing
 - [x] I have written automated positive tests
 - [ ] I have written automated negative tests
 - [ ] I have written zero automated tests but have poked around locally to verify proper functionality
 - [ ] The jira ticket has acceptance criteria and Kiara has in the information she needs to test my changes
 
## Release
 - [x] These changes require no special release procedures--just code!
 - [ ] Releasing these changes requires special handling
   - [ ] I have documented the special procedures in the release plan document
   - [ ] The special procedures require minimal downtime
   - [ ] The special procedures require unavoidable, extended downtime for:
     - [ ] Participant UX
     - [ ] Study staff

### CODE REVIEW

File: pepper-apis/src/test/java/org/broadinstitute/ddp/route/UserRegistrationRouteTest.java [hunk 1]
```diff
@@ -86,10 +86,14 @@
 import org.junit.Before;
 import org.junit.BeforeClass;
 import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 import org.testcontainers.shaded.org.apache.http.HttpStatus;
 
 public class UserRegistrationRouteTest extends IntegrationTestSuite.TestCase {
 
+    private static final Logger LOG = LoggerFactory.getLogger(StudyGovernanceDao.class);
```
Comment (yufengwng) [inline pepper-apis/src/test/java/org/broadinstitute/ddp/route/UserRegistrationRouteTest.java]: Typo?

**Comment by zyme** [comment]
please run tests

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/StudyGovernanceDao.java [hunk 2]
```diff
@@ -64,6 +68,20 @@ default GovernancePolicy createPolicy(GovernancePolicy policy) {
     @UseRowReducer(PolicyWithAgeOfMajorityRuleReducer.class)
     Optional<GovernancePolicy> findPolicyByStudyGuid(@Bind("studyGuid") String studyGuid);
 
+    /**
+     * Deletes the policy and any related rules.
+     */
+    default void deletePolicy(long policyId) {
+        var studyGovernanceSql = getStudyGovernanceSql();
+        int numRulesDeleted = studyGovernanceSql.deleteRulesForPolicy(policyId);
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/StudyGovernanceDao.java]: There's cascade delete in `age_of_majority_rule` so deleting the policy should delete the rules. I recommend not doing this extra work, unless we wanted to log how many rules were deleted.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/route/UserRegistrationRoute.java [hunk 3]
```diff
@@ -152,17 +154,34 @@ public Object handle(Request request, Response response, UserRegistrationPayload
                     throw new DDPException("There is already an account-bearing user for invitation " + invitationGuid);
                 }
 
-                // todo  arz check ageup status and process trigger events
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/route/UserRegistrationRoute.java:155]: Can we add a `todo` so we come back to add triggering events?

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/StudyGovernanceDao.java (hunk 2 above)
Comment (zyme) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/StudyGovernanceDao.java]: Sweet!

**Review by yufengwng — DISMISSED** [review]
👍

**Comment by zyme** [comment]
run tests

**Comment by zyme** [comment]
run tests

File: pepper-apis/src/test/java/org/broadinstitute/ddp/route/UserRegistrationRouteTest.java [hunk 4]
```diff
@@ -839,12 +854,27 @@ public void testAccountInvitation() {
...
-                        studyId, user.get().getId(), "test+" + System.currentTimeMillis() + "@datadonationplatform.org"));
+                        tempStudyId, user.get().getId(), "test+" + System.currentTimeMillis() + "@datadonationplatform.org"));
 
                 handle.attach(InvitationDao.class).clearDates(invitation.get().getInvitationGuid());
                 userDao.updateAuth0UserId(user.get().getGuid(), null);
+
+                UserProfileDto profile = new UserProfileDto(user.get().getId(), "Foo", "Bar", null,
+                        LocalDate.of(1953, 9, 18), null, null, null);
+
+                handle.attach(JdbiProfile.class).insert(profile);
```
Comment (yufengwng) [inline pepper-apis/src/test/java/org/broadinstitute/ddp/route/UserRegistrationRouteTest.java]: Looks like this line is causing the tests to fail? Jenkins is saying duplicate profile so I assume there's already an existing profile. This test is using the baked in test user, so I believe that user already has a profile.

Comment (zyme) [reply pepper-apis/src/test/java/org/broadinstitute/ddp/route/UserRegistrationRouteTest.java]: Yes, I've  been fixing this locally.  I was running the test in isolation and it was  happy, but then with all the other  tests, not so much...

**Pull request merged**